File: sdk/go/traceparent.go
package observr

import (
	"errors"
	"fmt"
	"strings"
)

// ParseTraceparent parses a W3C traceparent header.
// The returned Span represents the remote parent: its SpanID is the parent-id
// field from the header, so ObservrClient.Span() will correctly link to it.
func ParseTraceparent(header string) (*Span, error) {
	parts := strings.Split(header, "-")
	if len(parts) != 4 || parts[0] != "00" {
		return nil, errors.New("observr: invalid traceparent header")
	}
	if !isLowercaseHex(parts[1], 32) || !isLowercaseHex(parts[2], 16) {
		return nil, errors.New("observr: invalid traceparent header")
	}
	return &Span{
		TraceID: parts[1],
		SpanID:  parts[2],
	}, nil
}

// FormatTraceparent formats a Span as a W3C traceparent header value.
func FormatTraceparent(s *Span) string {
	return fmt.Sprintf("00-%s-%s-01", s.TraceID, s.SpanID)
}

func isLowercaseHex(s string, length int) bool {
	if len(s) != length {
		return false
	}
	for _, c := range s {
		if !((c >= '0' && c <= '9') || (c >= 'a' && c <= 'f')) {
			return false
		}
	}
	return true
}

File: sdk/node/src/traceparent.ts
// sdk/node/src/traceparent.ts
const LOWER_HEX = /^[0-9a-f]+$/;

/**
 * Parse a W3C traceparent header.
 * Returns { traceId, parentId } on success, null for any invalid input.
 * parentId should be used as parent_span_id so the causal chain links correctly.
 */
export function parseTraceparent(
  header: string
): { traceId: string; parentId: string } | null {
  const parts = header.split("-");
  if (parts.length !== 4 || parts[0] !== "00") return null;
  const [, traceId, parentId, flags] = parts;
  if (traceId.length !== 32 || parentId.length !== 16 || flags.length !== 2)
    return null;
  if (
    !LOWER_HEX.test(traceId) ||
    !LOWER_HEX.test(parentId) ||
    !LOWER_HEX.test(flags)
  )
    return null;
  if (/^0+$/.test(traceId) || /^0+$/.test(parentId)) return null;
  return { traceId, parentId };
}

/** Format a W3C traceparent header value. */
export function formatTraceparent(traceId: string, spanId: string): string {
  return `00-${traceId}-${spanId}-01`;
}
